modules/streams/arrow-streams/src/main/kotlin/arrow/streams/internal/FreeC.kt

```diff
@@ -0,0 +1,432 @@
+package arrow.streams.internal
+
+import arrow.Kind
+import arrow.core.*
+import arrow.effects.ExitCase
+import arrow.streams.CompositeFailure
+import arrow.streams.internal.FreeC.Result
+import arrow.typeclasses.MonadError
+
+//TODO temporary here until moved instances to separate module
+class ForFreeC private constructor() {
+  companion object
+}
+typealias FreeCOf<F, R> = arrow.Kind2<ForFreeC, F, R>
+typealias FreeCPartialOf<F> = arrow.Kind<ForFreeC, F>
+
+@Suppress("UNCHECKED_CAST", "NOTHING_TO_INLINE")
+inline fun <F, R> FreeCOf<F, R>.fix(): FreeC<F, R> =
+  this as FreeC<F, R>
+
+/**
+ * Free Monad with Catch (and Interruption).
+ *
+ * [FreeC] provides mechanism for ensuring stack safety and capturing any exceptions that may arise during computation.
+ *
+ * Furthermore, it may capture Interruption of the evaluation, although [FreeC] itself does not have any
+ * interruptible behaviour per se.
+ *
+ * Interruption cause may be captured in [FreeC.Interrupted] and allows user to pass along any information relevant
+ * to interpreter.
+ *
+ * Typically the [FreeC] user provides interpretation of FreeC in form of [ViewL] structure through the [ViewL.fold] function,
+ * that allows to step FreeC via series of Results ([FreeC.Pure], [FreeC.Fail] and [FreeC.Interrupted]) and FreeC step ([ViewL.View]).
+ */
+sealed class FreeC<F, out R> : FreeCOf<F, R> {
+
+  @Suppress("UNCHECKED_CAST", "ThrowRuntimeException")
+  fun <R2> flatMap(f: (R) -> FreeCOf<F, R2>): FreeC<F, R2> = FreeC.FlatMapped(this) { free: FreeC.Result<R> ->
+    when (free) {
+      is Pure<*, R> -> {
+        try {
+          f(free.r)
+        } catch (t: Throwable) {
+          FreeC.Fail<F, R2>(t)
+        }
+      }
+      is Fail<*, R> -> free
+      is Interrupted<*, R, *> -> free
+      else -> throw AssertionError("Unreachable")
+    } as FreeC<F, R2>
+  }
+
+  @Suppress("UNCHECKED_CAST", "ThrowRuntimeException")
+  fun <R2> map(f: (R) -> R2): FreeC<F, R2> = FlatMapped(this) { free ->
+    when (free) {
+      is Pure<*, R> -> {
+        try {
+          FreeC.Pure<F, R2>(f(free.r))
+        } catch (t: Throwable) {
+          FreeC.Fail<F, R2>(t)
+        }
+      }
+      is Fail<*, R> -> free
+      is Interrupted<*, R, *> -> free
+      else -> throw AssertionError("Unreachable")
```

Comment:
In the event that this is ever thrown we should indicate in the exception message that this is potentially caused by a bug and provide a link to report it in the arrow issue tracker.